src/Components/CustomerDataPersistor/CustomerDataPersistor.php

```diff
@@ -0,0 +1,125 @@
+    {
+        $this->saveOrderData($order, $dataBag, $context);
+
+        if ($this->configService->getBool('PayonePayment.settings.saveUserEnteredDataToCustomer')) {
+            $customer = $this->getCustomer($order, $context);
+            if ($customer instanceof CustomerEntity) {
+                $this->saveCustomerData($customer, $dataBag, $context);
+                $customerAddress = $this->getCustomerAddress($customer, $context);
+                if ($customerAddress instanceof CustomerAddressEntity) {
+                    $this->saveCustomerAddressData($customerAddress, $dataBag, $context);
+                }
+            }
+        }
+    }
+
+    protected function saveCustomerData(CustomerEntity $customer, RequestDataBag $dataBag, Context $context): void
+    {
+        $birthday = $dataBag->get(RequestConstants::BIRTHDAY);
+        $birthday = \is_string($birthday) ? \DateTime::createFromFormat('Y-m-d', $birthday) ?: null : null;
+
+        if ($birthday instanceof \DateTime && $birthday->getTimestamp() !== $customer->getBirthday()?->getTimestamp()) {
+            $this->customerRepository->update(
+                [
+                    [
+                        'id' => $customer->getId(),
+                        'birthday' => $birthday,
+                    ],
+                ],
+                $context
+            );
+        }
+    }
+
+    protected function saveCustomerAddressData(CustomerAddressEntity $customerAddress, RequestDataBag $dataBag, Context $context): void
+    {
+        $phoneNumber = $dataBag->get(RequestConstants::PHONE);
+
+        if ($phoneNumber !== $customerAddress->getPhoneNumber()) {
+            $this->customerAddressRepository->update(
+                [
+                    [
+                        'id' => $customerAddress->getId(),
+                        'phoneNumber' => $phoneNumber,
+                    ],
+                ],
+                $context
+            );
+        }
+    }
+
+    protected function saveOrderData(OrderEntity $orderEntity, RequestDataBag $dataBag, Context $context): void
+    {
+        $phoneNumber = $dataBag->get(RequestConstants::PHONE);
+
+        $this->orderAddressRepository->update(
+            [
+                [
+                    'id' => $orderEntity->getBillingAddressId(),
+                    'phoneNumber' => $phoneNumber,
+                ],
+            ],
+            $context
+        );
+    }
+
+    private function getCustomer(OrderEntity $order, Context $context): ?CustomerEntity
+    {
+        $customer = $order->getOrderCustomer()?->getCustomer();
+        if ($customer instanceof CustomerEntity) {
+            return $customer;
+        }
+        $customerId = $order->getOrderCustomer()?->getCustomerId();
+        if (empty($customerId)) {
+            return null;
+        }
+
+        /** @var CustomerEntity|null $customer */
+        $customer = $this->customerRepository->search(new Criteria([$customerId]), $context)->first();
```

Comment:
If you get to this point, it would make sense to add the “defaultBillingAddress” association here, as you need it in the “getCustomerAddress” function.

Reply:
good point! Done.